File: client/src/main/scala/stasis/client/tracking/ServerTracker.scala
package stasis.client.tracking

import java.time.Instant

import scala.concurrent.Future

import org.apache.pekko.NotUsed
import org.apache.pekko.stream.scaladsl.Source

trait ServerTracker extends ServerTracker.View {
  def reachable(server: String): Unit
  def unreachable(server: String): Unit
}

object ServerTracker {
  final case class ServerState(reachable: Boolean, timestamp: Instant)

  trait View {
    def state: Future[Map[String, ServerState]]
    def updates(server: String): Source[ServerState, NotUsed]
  }
}

File: client-ios/lib/Sources/StasisClientLib/Tracking/ServerTracker.swift
import Foundation

public protocol ServerTracker: Sendable {
    func reachable(server: String) async
    func unreachable(server: String) async
}

public struct ServerState: Sendable, Equatable, Hashable {
    public let reachable: Bool
    public let timestamp: Date

    public init(reachable: Bool, timestamp: Date) {
        self.reachable = reachable
        self.timestamp = timestamp
    }
}
